Frontier Mail

adeinitas opened the pull request:
## About the PR
<!-- What did you change in this PR? -->
Fixes #1031, sort of.
Ports changes made to mail that have been made by Frontier, that clean up the code of the mail system nicely by tweaking C#. Needs to be tested a bit more, but seems to run smoothly for the most part.
**What this PR does:**
- Adds the Mail RPDS, and gives the courierdrobe two of it. Couriers now have access to the mail gun
- Adds a shit ton of Frontier Mail, overhauls existing Nyano Mail, localizes mail.yml, all of the files for department mail yaml and mailDeliveries.yml to DeltaV folders. Obliterates a bunch of nyano files as a result, and rebalances the mail table.
- creates frontierMigrations.yml because Velcro mapped a mail prototype (WHY?!)

**Some things that need to be done later, out of the scope of this PR maybe:**
- Localization of mail papers
- Just make more mail tbh. I need people to be sent more fucked up things. If you have any ideas for mail that can provoke RP i'm open
## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
Mail is old, and Frontier made some nice changes to mail, and I want the courier to be able to shoot people with the mail china lake.
Furthermore, mail can be used to combat people meta-ing certain items like syndie clothes if you put those in the mail. 
## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Changes MailSystem C#, creates an entirely new mail table that shares a load of stuff with Frontier, creates frontierMigrations. Mostly C#, YAML, and I guess map changes because frontierMigrations.

## Media
![mailPR1RPDS](https://github.com/user-attachments/assets/7e061f9f-764d-4220-98f3-2acf9e3687f5)
![mailpr2smallenvelope](https://github.com/user-attachments/assets/dbad9226-d09f-45cc-880f-29bc8034e03d)
![mailPR3largemail](https://github.com/user-attachments/assets/9f916cdd-2bca-4fa3-b057-248c1e9a64bc)
![variousmailcontents](https://github.com/user-attachments/assets/695e5886-7a60-4184-9c69-96285a9a21ac)

## Requirements
<!-- 
Due to influx of PR's we require to ensure that PR's are following the correct guidelines.

Please take a moment to read these if its your first time.

Check the boxes below to confirm that you have in fact seen these (put an X in the brackets, like [X]):
-->
- [X] I have read and I am following the [Pull Request Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html). I understand that not doing so may get my pr closed at maintainer’s discretion
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->
:cl: Whatstone, dvir001, ErhardSteinhauer, Adeinitas
- tweak: New ads are in the mail and various mail has been expanded upon.
- add: You may find larger parcels in the mail.
- add: The mail RPDS is now in the courierdrobe - shoot your friends (and enemies) with the mail gun today!!!
<!--
Make sure to take this Changelog template out of the comment block in order for it to show up. Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog.
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

ErhardSteinhauer committed:
Mail Handheld Rapid Parcel Delivery System (#974)

* Mail RPDS

* removed sizes, more items go in capsules

* fixed error

* capsule resprite, added it to service techfab

* Update after upstream merge

* glass

* Update frontier.yml

* tag fix

---------

Co-authored-by: Dvir <dvirf01@gmail.com>

dvir001 committed:
Mail Buff (#1473)

* MailUp

* Update base_mail.yml

* Update base_mail.yml

* Update MailTeleporterComponent.cs

* Update MailComponent.cs

* Update base_mail.yml

* Update base_mail.yml

* Update Resources/Prototypes/Nyanotrasen/Entities/Objects/Specific/Mail/base_mail.yml

Sure

Co-authored-by: Whatstone <166147148+whatston3@users.noreply.github.com>

---------

Co-authored-by: Whatstone <166147148+whatston3@users.noreply.github.com>

whatston3 committed:
Mail table rework (#1477)

* Mail table, system for smoke prank

* Format mail messages

* DelayedItemSystem cleanup, more mail table entries

* More mail table entries

* Remove hoverbike from mail pool

* Add large mail.

FIXME: mail_large.rsi has no copyright info

* Default mail isLarge to false, comment cleanup

* More mail types

* Add optional large mail flag to admin mailto cmd

* mail_large copyright

* admin cmd fix, new mail types, const cleanup

* Mail: new components, weighting, a few new items

* Fix merge conflict, add placeholder pipebomb mail

* Format mail text, separate sword mail from knives

* Mail: fix exp. welder ID, split up Dan's cigs

* Fourth muffin, decrease captain's sabre chance

* yaml fixes

* Add ShowJobIcons component to mail hud

* Reorganize mail items, add build-a-buddy mail

* Build-a-Buddy fixes, slime & vulp versions, text

* Reptillian->Reptilian

* More signatures, label necrosol bottle, bigger emp

* lowercase i

* extra premium cigars, premium liquors

* Cleanup

* Fixups And Edits

* Remove kendo hakama, jabroni comment, fix cigars

* Platinum cigars, fix sprites

* fix premium absinthe ID

* Remove cyberpen, add BibleUserImplanter, ATV mail

Also adjusts weights for TacticalMaid (missing a zero), Restraints (cut
in half)

* kendo mail order, more mail comments

* Remove Nyano mail lists & parcels, move into _NF/

* True to true, cigars aren't fragile

---------

Co-authored-by: Dvir <dvirf01@gmail.com>
Co-authored-by: Whatstone <whatstone3@gmail.com>

Commit:
Does all the under the hood changes to mail components before I EVEN start thinking about MailSystem

dvir001 commented:
You probably missing some of the mail related PR's from our side.
Also we just had another PR to fix some issues with the mail app code and port it.

Commit:
Merge branch 'master' into The-Mail-Refactor

ghost commented:
> You probably missing some of the mail related PR's from our side. Also we just had another PR to fix some issues with the mail app code and port it.

I'm draft PRing this because I intend on working on it again. It's 2 months old. I gave up on it because the frontier mail table PR closed and I was having trouble cherrypicking specific shit from that.
I am missing the one that made a number of fixes to mail because it was part of the mail table ( that's frontier-station-14's #1477 ) and then I need to track down the rest of the mail fixes, nuke the frontier-specific calculations that account for addressing to the multiple ships (since this is only one station, i think), and then create the DV mail table.

Commit:
Kills accidentally cherrypicked platinum cig, makes mail be sorted into 5 ymls again

Commit:
THE MAIL TABLE!!!!! AHHHHHHHHHHHHHHHHHHHH

Commit:
NUKES two UNNECCESSARY RANDOMLY PASTED BRACKETS

FUCK YOU BUILD ERROR !!! EAT SHIT

Commit:
a semicolon for good measure

Adei versus the one silly

Commit:
Fuck is this? Why is it in my mail?

Commit:
Pro forgot values for two of the engineering mails

:skull:

Commit:
First pass to fix various mail bugs + editing of papers

Commit:
obliterates a random S in a C# file, also returns the mail teleporter to how it was

Commit:
Merge branch 'master' into The-Mail-Refactor

ghost commented:
Okay, I think this is complete enough to be opened and reviewed, but not merged **_whatsoever_**. The one thing it seems to be continuing to throw up whenever I open my devenv and test things is the platinum cigar's RSI, even though I exploded the cigar and it's gone. Look in at your own peril - I definitely need more eyes on this because there's just so many mistakes and there may or may not be things I either forgot to add from the frontier mail PR or things that break when opened.

Commit:
Nukes the platinum cigar's YAML

Since there's no intent for it to spawn in the mail, might as well.

Commit:
Merge branch 'The-Mail-Refactor' of https://github.com/adeinitas/Delta-v into The-Mail-Refactor

MilonPL commented on Resources/Locale/en-US/Mail/mail.ftl:
```diff
@@ -28,3 +29,8 @@ command-mailto-success = Success! Mail parcel has been queued for next teleport
+# Frontier: mailtestbulk
+command-mailtestbulk = Sends one of each type of parcel to a given mail teleporter.  Implicitly calls mailnow.
+command-mailtestbulk-help = Usage: {$command} <teleporter_id>
+command-mailtestbulk-success = Success! All mail teleporters will be delivering another round of mail soon.
```
Move to frontier ftl file.

ghost commented:
Wow, the linter has came for me, and it's throwing fucking hard punches. Jesus christ.

MilonPL commented on Content.Server/Nyanotrasen/Mail/Components/MailTeleporterComponent.cs:
```diff
@@ -28,7 +28,7 @@ public sealed partial class MailTeleporterComponent : Component
         /// teleporter can deliver.
         /// </summary>
         [DataField("mailPool")]
```
String key not needed.

MilonPL commented on Content.Server/Nyanotrasen/Mail/Components/MailTeleporterComponent.cs:
```diff
@@ -104,5 +104,19 @@ public sealed partial class MailTeleporterComponent : Component
         /// </summary>
         [DataField("priorityMalus")]
```
String key not needed.

MilonPL commented on Content.Server/Nyanotrasen/Mail/Components/MailTeleporterComponent.cs:
```diff
@@ -104,5 +104,19 @@ public sealed partial class MailTeleporterComponent : Component
+        /// <summary>
+        /// What's the bonus for delivering a large package intact?
+        /// </summary>
+        [DataField("largeBonus")]
```
String key not needed.

MilonPL commented on Content.Server/Nyanotrasen/Mail/Components/MailTeleporterComponent.cs:
```diff
@@ -104,5 +104,19 @@ public sealed partial class MailTeleporterComponent : Component
+        /// <summary>
+        /// What's the malus for failing to deliver a large package?
+        /// </summary>
+        [DataField("largeMalus")]
```
String key not needed.

MilonPL commented on Content.Server/Nyanotrasen/Mail/MailCommands.cs:
```diff
@@ -104,6 +115,7 @@ public async void Execute(IConsoleShell shell, string argStr, string[] args)
 
         mailComponent.IsFragile = isFragile;
         mailComponent.IsPriority = isPriority;
+        mailComponent.IsLarge = isLarge;
```
Nyano file so it's technically fine but missing a comment here.

MilonPL commented on Resources/Locale/en-US/Mail/mail.ftl:
Entire folder should probably be moved to nyanotrasen, no idea why it's here.

MilonPL commented on Resources/Prototypes/Nyanotrasen/Catalog/Fills/Vending/Inventories/maildrobe.yml:
```diff
@@ -1,8 +1,10 @@
   id: MailDrobeInventory
   startingInventory:
-    ClothingUniformMailCarrier: 2
-    ClothingUniformSkirtMailCarrier: 2
```
No need to change what line it's on.

Commit:
Milon's Review

Does some shit to the mail / courierdrobe and gets rid of string keys. Thanks Milon

MilonPL commented on Resources/Locale/en-US/nyanotrasen/Mail/mail.ftl:
```diff
@@ -28,3 +29,8 @@ command-mailto-success = Success! Mail parcel has been queued for next teleport
+# Frontier: mailtestbulk
+command-mailtestbulk = Sends one of each type of parcel to a given mail teleporter.  Implicitly calls mailnow.
+command-mailtestbulk-help = Usage: {$command} <teleporter_id>
+command-mailtestbulk-success = Success! All mail teleporters will be delivering another round of mail soon.
```
Duplicate of the frontier mail.ftl

MilonPL commented on Resources/Prototypes/Nyanotrasen/Catalog/Fills/Vending/Inventories/maildrobe.yml:
```diff
@@ -1,8 +1,12 @@
-    ClothingUniformMailCarrier: 2
-    ClothingUniformSkirtMailCarrier: 2
+    ClothingUniformJumpsuitMailCarrier: 2
+    ClothingUniformJumpskirtMailCarrier: 2
```
Not a valid prototype because it's from NF + duplicate of the lines below it

MilonPL commented on Resources/Prototypes/Nyanotrasen/Catalog/Fills/Vending/Inventories/maildrobe.yml:
```diff
@@ -1,8 +1,12 @@
+    WeaponMailLake: 1 # Frontier
+    BoxMailCapsulePrimed: 2 # Frontier
+    ClothingUniformJumpsuitMailCarrier: 2
+    ClothingUniformJumpskirtMailCarrier: 2
```
?

MilonPL reviewed (changes requested):
Haven't looked at the yaml mail tables but I trust these are fine.

Commit:
Milon's Review 2, Electric Boogaloo

Commit:
Various YAML fixes as I try to satisfy the Linter and fail

Commit:
Fixes Mail.yml / If I get rid of the EMP china lake does it fix things? Let's find out!

Commit:
Cleans up more C# at Milon's behest

Commit:
TORTUGAAAAAAAAAAAAAAAAAAAA!!!!

Why is this mapped in?

ghost commented:
I guess it's calling Velcro now because I had to touch that tortuga mail prototype - that's just a thing of it being compatible with this port.

MilonPL commented on Resources/Maps/tortuga.yml:
PRO use MIGRATIONS instead of modifying the map
In deltaMigrations.yml add MailPAI: MailNFPAI

MilonPL commented on Content.Server/_NF/Mail/DelayedItemComponent.cs:
```diff
@@ -0,0 +1,16 @@
+        /// <summary>
+        /// The entity to replace this when opened or dropped.
+        /// </summary>
+        [DataField, ViewVariables(VVAccess.ReadWrite)]
```
ReadWrite not needed.

Commit:
Revert "TORTUGAAAAAAAAAAAAAAAAAAAA!!!!"

This reverts commit b1d415a7be6e5193fc77424b2847ff7babdf6cfa.

Commit:
Migrations

Commit:
Does a bit more fixing

DangerRevolution commented on Content.Server/Nyanotrasen/Mail/Components/MailComponent.cs:
```diff
@@ -58,43 +55,51 @@ public sealed partial class MailComponent : SharedMailComponent
+        [DataField]
         public bool IsPriority = false;
 
+        // Code imported from New Frontiers, licensed under MIT: large mail
```
I see we're already preparing for the Wizden AGPL nuke

MilonPL replied:
> I see we're already preparing for the Wizden AGPL nuke

Wizden is NOT going to take any action against AGPL servers is that PJB said. Don't spread misinformation.

DangerRevolution commented on Resources/Locale/en-US/_NF/mail/mail.ftl:
```diff
@@ -0,0 +1,9 @@
+mail-large-desc-far = A large package.
+mail-large-desc-close = A large package addressed to {CAPITALIZE($name)}, {$job}. Last known location: {$station}.
+
+# Frontier: mailtestbulk
```
```suggestion
```

.ftl doesn't support comments and this causes one of the 2000+ warnings to pop up when the game builds

MilonPL replied:
> .ftl doesn't support comments and this causes one of the 2000+ warnings to pop up when the game builds

It does, it's just that people can't format them properly. https://projectfluent.org/fluent/guide/comments.html

DangerRevolution commented on Resources/Locale/en-US/nyanotrasen/Mail/mail.ftl:
```diff
@@ -8,15 +8,16 @@ mail-desc-priority = The anti-tamper lock's [color=yellow]yellow priority tape[/
 
 command-mailto-description = Queue a parcel to be delivered to an entity. Example usage: `mailto 1234 5678 false false`. The target container's contents will be transferred to an actual mail parcel.
-command-mailto-help = Usage: {$command} <recipient entityUid> <container entityUid> [is-fragile: true or false] [is-priority: true or false]
+# Frontier: add is-large description
```
```suggestion
```


Same as above

DangerRevolution commented on Resources/Migrations/deltaMigrations.yml:
```diff
@@ -122,3 +122,6 @@ FoodMealSoftTaco: FoodTacoShell
 BoxMagazinePistolSubMachineGunRubber: MagazineBoxPistolRubber
+
+# 2024-08-22 - Frontier Mail - Add more to these when they come up as mapped
+MailPAI: MailNFPAI
```
```suggestion
```
No one should have ever mapped a mail-based pAI; their fault if they have.

Also, frontier migrations should be in frontierMigrations.yml

DangerRevolution commented on Resources/Prototypes/Nyanotrasen/Entities/Objects/Specific/Mail/base_mail.yml:
```diff
@@ -93,6 +95,16 @@
       types:
         Blunt: 10
+  - type: CargoSellBlacklist # Frontier
+  - type: Food # Frontier - Moth food
```
Why can moths eat paper?

Commit:
Update Resources/Locale/en-US/_NF/mail/mail.ftl

Co-authored-by: Danger Revolution! <142105406+DangerRevolution@users.noreply.github.com>
Signed-off-by: Adeinitas <147965189+adeinitas@users.noreply.github.com>

Commit:
Update Resources/Locale/en-US/nyanotrasen/Mail/mail.ftl

Co-authored-by: Danger Revolution! <142105406+DangerRevolution@users.noreply.github.com>
Signed-off-by: Adeinitas <147965189+adeinitas@users.noreply.github.com>

Commit:
Fixes Mail Capsule / Creates frontierMigrations

Blame Velcro for frontierMigrations.

Commit:
Merge branch 'The-Mail-Refactor' of https://github.com/adeinitas/Delta-v into The-Mail-Refactor

Commit:
Update Resources/Migrations/deltaMigrations.yml

Co-authored-by: Danger Revolution! <142105406+DangerRevolution@users.noreply.github.com>
Signed-off-by: Adeinitas <147965189+adeinitas@users.noreply.github.com>

Commit:
Rewords that one fucking comment I made 2 months ago about MIT / AGPL

Commit:
Merge branch 'The-Mail-Refactor' of https://github.com/adeinitas/Delta-v into The-Mail-Refactor

Commit:
Just gives the delayed smoke prank and delayed EMP effects the instant effect sprites

Commit:
Where the fuck are these random curly braces from??

Commit:
An attempt to satisfy the Yaml Linter

Commit:
Revert "Update Resources/Locale/en-US/nyanotrasen/Mail/mail.ftl"

This reverts commit 0ee3869d76cc7633be5fd26d303221e79ab04c77.

Commit:
Revert "Update Resources/Locale/en-US/_NF/mail/mail.ftl"

This reverts commit 29129df9d66079c9df0e4b1ccf5818b0cb230234.

Commit:
Mail FTL fixes / I think I know what the linter wants

Commit:
will this sprite be valid pretty please?

ghost commented:
All checks have passed, and with that, I think I'm finally done. Sorry for the commit spam, now an actual maint can review this and merge.

Commit:
Does a first pass on noSpawn for all mails.

ANOTHER COMMIT FOR THE PILE.

MilonPL reviewed (changes requested):
`nospawn: true` is obsolete and will be completely removed soon, should be replaced with `categories: [ HideSpawnMenu ]`, which isn't needed because it's inhereted from BaseMail.

ghost commented:
> `nospawn: true` is obsolete and will be completely removed soon, should be replaced with `categories: [ HideSpawnMenu ]`, which isn't needed because it's inhereted from BaseMail.

FUCK

Commit:
Unfucks noSpawn from the mail prototypes

ghost commented:
Okay, now this is Really Actually the end of me commit spamming this, trust.

ghost commented:
Don't need Velcro's insight because it's not a direct map change, anymore. Velcro, you're still free to look though, but the adding you to the PR was accidental
I'd like insight from the carp man and the other yamltainer to see what they think of this!
I'd also like insight from the Game Directors, once they are able to review.

therealDLondon reviewed (approved):
I'm Fox, and this is my favourite PR on the Repo.

Tests pass, Milon nods, the fish man weeps silently somewhere
